# Pandas Basics 

In this exercise we'll do some analysis on a real-world dataset. We'll ask research questions and attempt to provide answers to those questions using pandas and seaborn.
<br/>
Unlike in the real-world, the dataset you'll be working with has been mostly cleaned-up. There is very little you'll have to do with the data to get it all nice and tidy.

Our dataset is a subset of the 2019 StackOverflow Developers' Survey. The columns are as follows:
- **Country**: respondent's country
- **Age**: respondent's age
- **Gender**: respondent's gender
- **YearsCodePro**: years respondent has been developing professionaly
- **OrgSize**: repondent's organization size - string
- **WorkWeekHrs**: average number of hours the respondesnt works weekly
- **ConevrtedComp**: repondent's yearly compensation (converted to USD)
- **LanguageWorkedWith**: programming languages the respondent has worked with

In [3]:
import pandas as pd
import numpy as np

In [4]:
df = pd.read_csv("so2019_survey_results_short.CLEAN.csv")

In [5]:
df.sample(10)

Country   Age Gender  YearsCodePro  OrgSize  WorkWeekHrs  \
12115               India  24.0    Man             2     <100         40.0   
12876              Canada  33.0    Man            10     <500         65.0   
46937      United Kingdom  52.0    Man            36      <20         35.0   
1918        United States  57.0    Man            33   <5,000         40.0   
43658       United States  21.0    Man             1  >10,000         40.0   
39013               Nepal  24.0    Man             3     <100         40.0   
41755              Brazil  36.0    Man            10     <500         40.0   
44118       United States  27.0    Man             5  >10,000         45.0   
18882  Russian Federation  35.0    Man             8  <10,000         40.0   
19003              Sweden  26.0    Man             2  <10,000         38.0   

       ConvertedComp                                 LanguageWorkedWith  
12115        41976.0                                HTML/CSS;JavaScript  
12876        95423.0                        C++;HTML/CSS;JavaScript;SQL  
46937        65424.0                          Bash/Shell/PowerShell;C++  
1918        135000.0                                      SQL;Other(s):  
43658       120000.0                   C;C++;Go;Java;Kotlin;Python;Rust  
39013         4224.0  Assembly;Bash/Shell/PowerShell;C;C++;C#;Go;HTM...  
41755        48000.0                         HTML/CSS;JavaScript;Python  
44118        86250.0                                    C;C#;Python;SQL  
18882        35196.0               Bash/Shell/PowerShell;C++;Python;SQL  
19003        45996.0           Bash/Shell/PowerShell;C++;Go;Python;Ruby

In [6]:
df.loc[df.Country == 'France', ['Age', 'WorkWeekHrs']]

Age  WorkWeekHrs
32     31.0         40.0
120    40.0         37.0
155    23.0         45.0
198    35.0         40.0
207    23.0         39.0
...     ...          ...
51477  37.0         38.0
51487  23.0         35.0
51527  27.0         40.0
51560  22.0         70.0
51562  37.0         37.0

[1563 rows x 2 columns]

#### 1. Let's do a little warm-up:

1.1 Diaply 10 random rows.

In [7]:
df.sample(10)

Country   Age Gender  YearsCodePro        OrgSize  \
3717        United States  30.0    Man             1           <100   
17264       United States  33.0    Man            10           <100   
5180        United States  24.0    Man             2        >10,000   
27281       United States  35.0    Man            14        >10,000   
27703              Uganda  22.0    Man             1         <5,000   
41311  Russian Federation  33.0    Man             2           <100   
45735             Austria  38.0    Man            15  self-employed   
16300       United States  28.0  Woman             6            <20   
17780               China  25.0    Man             3           <500   
48624      United Kingdom  33.0    Man            10           <100   

       WorkWeekHrs  ConvertedComp  \
3717          40.0        41600.0   
17264         55.0        80000.0   
5180          40.0        70000.0   
27281         40.0       160000.0   
27703         45.0         4596.0   
41311         45.0         1836.0   
45735         50.0        41244.0   
16300         35.0        95000.0   
17780         40.0        21348.0   
48624         40.0        52339.0   

                                      LanguageWorkedWith  
3717                      HTML/CSS;JavaScript;PHP;Python  
17264                Bash/Shell/PowerShell;C#;Python;SQL  
5180                                   C;C++;Java;Python  
27281  Bash/Shell/PowerShell;HTML/CSS;JavaScript;Ruby...  
27703                    HTML/CSS;Java;JavaScript;Python  
41311                                         JavaScript  
45735                    C#;HTML/CSS;Java;JavaScript;SQL  
16300                                        Java;Kotlin  
17780  Bash/Shell/PowerShell;C++;C#;Go;HTML/CSS;JavaS...  
48624  Bash/Shell/PowerShell;C;C#;HTML/CSS;JavaScript...

1.2 Display the DF's meta-data.

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51582 entries, 0 to 51581
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Country             51582 non-null  object 
 1   Age                 51582 non-null  float64
 2   Gender              51582 non-null  object 
 3   YearsCodePro        51582 non-null  int64  
 4   OrgSize             51582 non-null  object 
 5   WorkWeekHrs         51582 non-null  float64
 6   ConvertedComp       51582 non-null  float64
 7   LanguageWorkedWith  51427 non-null  object 
dtypes: float64(3), int64(1), object(4)
memory usage: 3.1+ MB


1.3 Extract the columns *Country* and *Age*. 

In [9]:
df[['Country', 'Age']]

Country   Age
0           Thailand  28.0
1      United States  22.0
2        New Zealand  23.0
3      United States  28.0
4            Germany  31.0
...              ...   ...
51577  United States  48.0
51578  United States  26.0
51579        Finland  34.0
51580        Austria  37.0
51581         Greece  33.0

[51582 rows x 2 columns]

1.4 Extract the columns *WorkWeekHrs* and *Age* where *Country* is France.

In [10]:
df.loc[df.Country == 'France', ['WorkWeekHrs', 'Age']]

WorkWeekHrs   Age
32            40.0  31.0
120           37.0  40.0
155           45.0  23.0
198           40.0  35.0
207           39.0  23.0
...            ...   ...
51477         38.0  37.0
51487         35.0  23.0
51527         40.0  27.0
51560         70.0  22.0
51562         37.0  37.0

[1563 rows x 2 columns]

In [11]:
df['Age'] >= 40

0        False
1        False
2        False
3        False
4        False
         ...  
51577     True
51578    False
51579    False
51580    False
51581    False
Name: Age, Length: 51582, dtype: bool

1.5 Display a list of countries (each country should show once).

In [12]:
df.Country.unique() # returns numpy of type array and not series

array(['Thailand', 'United States', 'New Zealand', 'Germany',
       'United Kingdom', 'Australia', 'Brazil', 'Lithuania', 'Israel',
       'South Africa', 'Switzerland', 'Argentina', 'Czech Republic',
       'Spain', 'Serbia', 'India', 'Poland', 'Sweden', 'China', 'France',
       'Netherlands', 'Philippines', 'Ireland', 'Pakistan', 'Austria',
       'Canada', 'Croatia', 'Russian Federation', 'Italy', 'Turkey',
       'Greece', 'Romania', 'Finland', 'Bulgaria', 'Viet Nam', 'Iran',
       'Hungary', 'Latvia', 'Hong Kong (S.A.R.)', 'United Arab Emirates',
       'Portugal', 'Nigeria', 'Norway', 'Nicaragua', 'Denmark', 'Japan',
       'Bangladesh', 'Mexico', 'Ecuador', 'Singapore', 'Ukraine',
       'South Korea', 'Slovenia', 'Estonia', 'Belgium', 'Mongolia',
       'Chile', 'Malaysia', 'Luxembourg', 'Dominican Republic', 'Peru',
       'Nepal', 'Colombia', 'Guatemala', 'Taiwan', 'Yemen', 'Indonesia',
       'Morocco', 'Sri Lanka', 'Slovakia', 'Kenya',
       'Venezuela, Bolivarian Repub

1.6 Display a list of countries (each country should show once), sorted alphabetical order.

In [13]:
df.Country.sort_values().unique()

array(['Afghanistan', 'Albania', 'Algeria', 'Andorra', 'Angola',
       'Argentina', 'Armenia', 'Australia', 'Austria', 'Azerbaijan',
       'Bahrain', 'Bangladesh', 'Belarus', 'Belgium', 'Bolivia',
       'Bosnia and Herzegovina', 'Botswana', 'Brazil',
       'Brunei Darussalam', 'Bulgaria', 'Burkina Faso', 'Burundi',
       'Cambodia', 'Cameroon', 'Canada', 'Chad', 'Chile', 'China',
       'Colombia', 'Congo, Republic of the...', 'Costa Rica', 'Croatia',
       'Cuba', 'Cyprus', 'Czech Republic', "Côte d'Ivoire",
       "Democratic People's Republic of Korea",
       'Democratic Republic of the Congo', 'Denmark', 'Djibouti',
       'Dominican Republic', 'Ecuador', 'Egypt', 'El Salvador', 'Estonia',
       'Ethiopia', 'Fiji', 'Finland', 'France', 'Georgia', 'Germany',
       'Ghana', 'Greece', 'Guatemala', 'Guinea', 'Haiti', 'Honduras',
       'Hong Kong (S.A.R.)', 'Hungary', 'Iceland', 'India', 'Indonesia',
       'Iran', 'Iraq', 'Ireland', 'Israel', 'Italy', 'Jamaica', 'Japan',
    

1.7 Convert the compensation (ConvertedComp) to thousands of USD (K), rounding the number to the nearest 1,000.
<br/>
For example: 20,628 --> 21

Hint: look-up the .round() method.

In [14]:
df2 = df
df2['USD_K_ConvertedComp'] = round(df.ConvertedComp / 1000).astype(int)
df2
# sugested solution:
# df.ConvertedComp = (df.ConvertedComp / 1000).round(0).astype("int")    

Country   Age Gender  YearsCodePro OrgSize  WorkWeekHrs  \
0           Thailand  28.0    Man             1    <500        40.00   
1      United States  22.0    Man             1    <500        80.00   
2        New Zealand  23.0    Man             4     <20        32.00   
3      United States  28.0    Man             8     <20        40.00   
4            Germany  31.0    Man             2    <500        40.00   
...              ...   ...    ...           ...     ...          ...   
51577  United States  48.0    Man            28   <1000        45.00   
51578  United States  26.0    Man             3    <100        40.00   
51579        Finland  34.0    Man             7    <100        37.75   
51580        Austria  37.0    Man             9    <500        39.00   
51581         Greece  33.0    Man             5    <100        48.00   

       ConvertedComp                                 LanguageWorkedWith  \
0             8820.0                                           HTML/CSS   
1            61000.0                                C;C++;C#;Python;SQL   
2            95179.0  Bash/Shell/PowerShell;C#;HTML/CSS;JavaScript;P...   
3            90000.0  Bash/Shell/PowerShell;HTML/CSS;JavaScript;PHP;...   
4            57060.0                                                C++   
...              ...                                                ...   
51577      2000000.0  Bash/Shell/PowerShell;C;Clojure;HTML/CSS;Java;...   
51578       130000.0               HTML/CSS;JavaScript;Scala;TypeScript   
51579        82488.0                   Bash/Shell/PowerShell;C++;Python   
51580        68745.0  Bash/Shell/PowerShell;Go;HTML/CSS;Java;JavaScr...   
51581        22915.0              Go;HTML/CSS;Java;JavaScript;Other(s):   

       USD_K_ConvertedComp  
0                        9  
1                       61  
2                       95  
3                       90  
4                       57  
...                    ...  
51577                 2000  
51578                  130  
51579                   82  
51580                   69  
51581                   23  

[51582 rows x 9 columns]

1.8 Which columns have missing values? How many?

In [15]:
#  any() method returns one value for each column, True if ANY value in that column is True, otherwise False.
# col_with_nulls = df.columns[df.isna().any() == True]
# df[col_with_nulls].isna().sum()

# Sugested solution
df.isna().sum()

Country                  0
Age                      0
Gender                   0
YearsCodePro             0
OrgSize                  0
WorkWeekHrs              0
ConvertedComp            0
LanguageWorkedWith     155
USD_K_ConvertedComp      0
dtype: int64

#### 2. Perform initial exploration of the data using ``describe``. Can you identify potential problems in the data?

In [16]:
df.describe()

Age  YearsCodePro   WorkWeekHrs  ConvertedComp  \
count  51582.000000  51582.000000  51582.000000   5.158200e+04   
mean      31.448180      8.053817     41.996582   1.271387e+05   
std        8.111043      7.185080     35.896054   2.832771e+05   
min        1.000000      1.000000      1.000000   0.000000e+00   
25%       26.000000      3.000000     40.000000   2.623200e+04   
50%       30.000000      6.000000     40.000000   5.728700e+04   
75%       35.000000     11.000000     43.000000   1.000000e+05   
max       99.000000     50.000000   4850.000000   2.000000e+06   

       USD_K_ConvertedComp  
count         51582.000000  
mean            127.130879  
std             283.277537  
min               0.000000  
25%              26.000000  
50%              57.000000  
75%             100.000000  
max            2000.000000

the potential problems:
1) outliers in Age column (min = 1, max = 99) suggest incorrect data 
2) outliers in YearsCodePro column (max = 50) suggest incorrect data 
3) outliers in WorkWeekHrs column (max = 4850) suggest incorrect data 
4) outliers in ConvertedComp column (max = 2.000000e+06) suggest incorrect data 

#### 3. Lets remove the outliers for *Age* and *WorkWeekHrs*. - not for home work

For now (until we learn more about data pre-processing), lets assume we'd like to discard of:
1. Respondents in their pre-teens (<15) and those older than 80
2. Respondents working less than 10 hours a week, or more than 120 hours a week

**Hints:**
- Write the mask's condition(s) on their own, and see that you get what you expected
- DF's ``drop`` method accepts indices only. You will have to use the ``.index`` property on the DF filtered by the mask (see the tutorial for an example)

3.1 Drop rows in which age < 15 or age > 80:

In [17]:
#drop works per index
df[(df.Age < 15) | (df.Age > 80)].index

Int64Index([ 5621, 11785, 16614, 16943, 17102, 17808, 22228, 24737, 25544,
            26619, 28992, 29387, 33700, 35321, 39625, 44022, 45729, 46728,
            48066, 49126, 49499, 51222],
           dtype='int64')

In [18]:
df.drop(df[(df.Age < 15) | (df.Age > 80)].index, inplace=True)

3.2 Drop rows in which weekly work hours < 10 or weekly work hours > 120

In [19]:
df.drop(df[(df.WorkWeekHrs < 10) | (df.WorkWeekHrs > 120)].index, inplace=True)

#### 4. Add an *ExpLevel* column categorizing experience years into four groups

0 to 3 years  --> junior<br/>
3 to 7 years  --> experienced<br/>
7 to 10 years --> senior<br/>
10+ years     --> master

In [20]:
df_exp_level = df
df_exp_level['ExpLevel'] = pd.cut(
    df['YearsCodePro'],
    bins=[0, 3, 7, 10, df['YearsCodePro'].max()],
    labels=['junior', 'experienced', 'senior', 'master']
)
df_exp_level
# sugested solution
# df["ExpLevel"] = pd.cut(df.YearsCodePro, bins=[0, 3, 7, 10, 99], labels=["junior", "experienced", "senior", "master"])

Country   Age Gender  YearsCodePro OrgSize  WorkWeekHrs  \
0           Thailand  28.0    Man             1    <500        40.00   
1      United States  22.0    Man             1    <500        80.00   
2        New Zealand  23.0    Man             4     <20        32.00   
3      United States  28.0    Man             8     <20        40.00   
4            Germany  31.0    Man             2    <500        40.00   
...              ...   ...    ...           ...     ...          ...   
51577  United States  48.0    Man            28   <1000        45.00   
51578  United States  26.0    Man             3    <100        40.00   
51579        Finland  34.0    Man             7    <100        37.75   
51580        Austria  37.0    Man             9    <500        39.00   
51581         Greece  33.0    Man             5    <100        48.00   

       ConvertedComp                                 LanguageWorkedWith  \
0             8820.0                                           HTML/CSS   
1            61000.0                                C;C++;C#;Python;SQL   
2            95179.0  Bash/Shell/PowerShell;C#;HTML/CSS;JavaScript;P...   
3            90000.0  Bash/Shell/PowerShell;HTML/CSS;JavaScript;PHP;...   
4            57060.0                                                C++   
...              ...                                                ...   
51577      2000000.0  Bash/Shell/PowerShell;C;Clojure;HTML/CSS;Java;...   
51578       130000.0               HTML/CSS;JavaScript;Scala;TypeScript   
51579        82488.0                   Bash/Shell/PowerShell;C++;Python   
51580        68745.0  Bash/Shell/PowerShell;Go;HTML/CSS;Java;JavaScr...   
51581        22915.0              Go;HTML/CSS;Java;JavaScript;Other(s):   

       USD_K_ConvertedComp     ExpLevel  
0                        9       junior  
1                       61       junior  
2                       95  experienced  
3                       90       senior  
4                       57       junior  
...                    ...          ...  
51577                 2000       master  
51578                  130       junior  
51579                   82  experienced  
51580                   69       senior  
51581                   23  experienced  

[49927 rows x 10 columns]

Now, kets do some analysis...

#### 5. Add a *WorkLevel* column categorizing weekly work hours into four groups

0 to 35 h/week  --> part-time<br/>
35 to 50 h/week  --> full-time<br/>
50 to 65 h/week --> dedicated<br/>
65+ h/week     --> workaholic

In [21]:
df_work_level = df
df_work_level['WorkLevel'] = pd.cut(
    df_work_level['WorkWeekHrs'],
    bins=[0, 35, 50, 65, df['WorkWeekHrs'].max()],
    labels=['part-time', 'full-time', 'dedicated', 'workaholic']
)
df_work_level
# sugested solution
# df["WorkLevel"] = pd.cut(df.WorkWeekHrs, bins=[0, 35, 50, 65, 200], labels=["part-time", "full-time", "dedicated", "workaholic"])

Country   Age Gender  YearsCodePro OrgSize  WorkWeekHrs  \
0           Thailand  28.0    Man             1    <500        40.00   
1      United States  22.0    Man             1    <500        80.00   
2        New Zealand  23.0    Man             4     <20        32.00   
3      United States  28.0    Man             8     <20        40.00   
4            Germany  31.0    Man             2    <500        40.00   
...              ...   ...    ...           ...     ...          ...   
51577  United States  48.0    Man            28   <1000        45.00   
51578  United States  26.0    Man             3    <100        40.00   
51579        Finland  34.0    Man             7    <100        37.75   
51580        Austria  37.0    Man             9    <500        39.00   
51581         Greece  33.0    Man             5    <100        48.00   

       ConvertedComp                                 LanguageWorkedWith  \
0             8820.0                                           HTML/CSS   
1            61000.0                                C;C++;C#;Python;SQL   
2            95179.0  Bash/Shell/PowerShell;C#;HTML/CSS;JavaScript;P...   
3            90000.0  Bash/Shell/PowerShell;HTML/CSS;JavaScript;PHP;...   
4            57060.0                                                C++   
...              ...                                                ...   
51577      2000000.0  Bash/Shell/PowerShell;C;Clojure;HTML/CSS;Java;...   
51578       130000.0               HTML/CSS;JavaScript;Scala;TypeScript   
51579        82488.0                   Bash/Shell/PowerShell;C++;Python   
51580        68745.0  Bash/Shell/PowerShell;Go;HTML/CSS;Java;JavaScr...   
51581        22915.0              Go;HTML/CSS;Java;JavaScript;Other(s):   

       USD_K_ConvertedComp     ExpLevel   WorkLevel  
0                        9       junior   full-time  
1                       61       junior  workaholic  
2                       95  experienced   part-time  
3                       90       senior   full-time  
4                       57       junior   full-time  
...                    ...          ...         ...  
51577                 2000       master   full-time  
51578                  130       junior   full-time  
51579                   82  experienced   full-time  
51580                   69       senior   full-time  
51581                   23  experienced   full-time  

[49927 rows x 11 columns]

#### 6. What are the top 5 countries with most respondents (name and number)?

An ordered DF of counts is okay.

**Hints:**
- What aggregation function needs to be invoked on the grouped DF?
- Sorting the counted values is in order.

In [22]:
df.head()

Country   Age Gender  YearsCodePro OrgSize  WorkWeekHrs  \
0       Thailand  28.0    Man             1    <500         40.0   
1  United States  22.0    Man             1    <500         80.0   
2    New Zealand  23.0    Man             4     <20         32.0   
3  United States  28.0    Man             8     <20         40.0   
4        Germany  31.0    Man             2    <500         40.0   

   ConvertedComp                                 LanguageWorkedWith  \
0         8820.0                                           HTML/CSS   
1        61000.0                                C;C++;C#;Python;SQL   
2        95179.0  Bash/Shell/PowerShell;C#;HTML/CSS;JavaScript;P...   
3        90000.0  Bash/Shell/PowerShell;HTML/CSS;JavaScript;PHP;...   
4        57060.0                                                C++   

   USD_K_ConvertedComp     ExpLevel   WorkLevel  
0                    9       junior   full-time  
1                   61       junior  workaholic  
2                   95  experienced   part-time  
3                   90       senior   full-time  
4                   57       junior   full-time

In [23]:
grpby_Country = df.groupby(['Country'])['Country'].count()
grpby_Country.sort_values(ascending=False)


Country
United States     13727
United Kingdom     3743
Germany            3457
India              3238
Canada             2056
                  ...  
Malawi                1
Seychelles            1
Liechtenstein         1
Togo                  1
Oman                  1
Name: Country, Length: 157, dtype: int64

In [24]:
grpby_Country.sort_values(ascending=False).iloc[0:5]

Country
United States     13727
United Kingdom     3743
Germany            3457
India              3238
Canada             2056
Name: Country, dtype: int64

In [25]:
grpby_Country.sort_values(ascending=False).head(5)

Country
United States     13727
United Kingdom     3743
Germany            3457
India              3238
Canada             2056
Name: Country, dtype: int64

In [72]:
df.Country.value_counts().head(5) # prefered solution

United States     13727
United Kingdom     3743
Germany            3457
India              3238
Canada             2056
Name: Country, dtype: int64

#### In the following questions, we'll focus on 7 countries: *United States*, *United Kingdom*, *Germany*, *France*, *Israel*, *Ukraine* and *India*. The assignment below will create a the needed DF for you. Just execute the cell...

In [26]:
df_countries = df[df.Country.isin(["United States", "United Kingdom", "Germany", "Israel", "Ukraine", "India"])]

In [27]:
df_countries.sample(5)

Country   Age Gender  YearsCodePro OrgSize  WorkWeekHrs  \
33285  United Kingdom  29.0    Man             3    <100         35.0   
24927         Germany  23.0    Man             2     <20         40.0   
39064           India  21.0    Man             1     <10         40.0   
32600   United States  22.0    Man             1    <500         40.0   
41485         Germany  31.0    Man            10    <100         42.0   

       ConvertedComp                                 LanguageWorkedWith  \
33285        48413.0                                               Ruby   
24927        42624.0                                HTML/CSS;JavaScript   
39064        26364.0  C;C++;Go;HTML/CSS;Java;JavaScript;Objective-C;...   
32600        60000.0  Bash/Shell/PowerShell;C++;C#;Erlang;HTML/CSS;J...   
41485        72182.0  Bash/Shell/PowerShell;C#;HTML/CSS;Java;JavaScr...   

       USD_K_ConvertedComp ExpLevel  WorkLevel  
33285                   48   junior  part-time  
24927                   43   junior  full-time  
39064                   26   junior  full-time  
32600                   60   junior  full-time  
41485                   72   senior  full-time

#### 7. What is the average salary and average weekly work hours as per experience level in the above subset of 7 countries?

In [28]:
df_grpby_ExpLevel = df_countries.groupby('ExpLevel').agg(             # agg function
    {
        'USD_K_ConvertedComp': pd.Series.mean,
        'WorkWeekHrs': pd.Series.mean
     }
)

df_grpby_ExpLevel.round(2)

USD_K_ConvertedComp  WorkWeekHrs
ExpLevel                                     
junior                    145.85        40.98
experienced               175.49        41.47
senior                    204.93        41.68
master                    218.90        42.09

In [73]:
# suggested solution
(
    df_countries
    .groupby(["Country", "ExpLevel"])
    [["WorkWeekHrs", "ConvertedComp"]]
    .mean()
)

WorkWeekHrs  ConvertedComp
Country        ExpLevel                               
Germany        junior         37.328804   72862.053261
               experienced    38.819465  106221.779520
               senior         40.715604  123507.203020
               master         40.522870  145424.714119
India          junior         44.824866   15000.164439
               experienced    43.849535   29779.192730
               senior         44.566879   44018.933121
               master         43.669388   71847.440816
Israel         junior         41.418103   59949.043103
               experienced    44.953125   94255.525000
               senior         44.618750  116194.912500
               master         44.716129  118870.722581
Ukraine        junior         39.857895   19151.552632
               experienced    39.869565   32585.659420
               senior         38.923077   40966.661538
               master         39.993671   61689.886076
United Kingdom junior         39.155742  126209.313008
               experienced    39.238273  165026.466068
               senior         39.487981  178369.798077
               master         39.684519  198023.570736
United States  junior         40.868229  228220.822396
               experienced    41.994679  254018.629930
               senior         42.109848  278240.958474
               master         42.934476  253598.857597

#### 8. What is the percentage of workaholic in Israel? In India?

In [39]:
df_countries_Israel_India = df[df.Country.isin(['Israel','India'])]


Country   Age Gender  YearsCodePro  OrgSize  WorkWeekHrs  ConvertedComp  \
11     Israel  23.0    Man             1      <20         50.0        26388.0   
39      India  29.0    Man             7      <20         70.0        11194.0   
45      India  27.0  Woman             7  >10,000         50.0        17491.0   
48      India  25.0    Man             3      <10         40.0         7696.0   
49      India  22.0    Man             1     <100         45.0       117540.0   
...       ...   ...    ...           ...      ...          ...            ...   
51443   India  24.0    Man             3     <100         46.0         4897.0   
51470   India  30.0    Man             4     <100         47.0         7052.0   
51482   India  28.0    Man             6   <5,000         40.0        37781.0   
51494   India  26.0    Man             4      <20         50.0         3360.0   
51545   India  26.0    Man             4   <5,000         40.0        13993.0   

                           LanguageWorkedWith  USD_K_ConvertedComp  \
11         HTML/CSS;JavaScript;PHP;TypeScript                   26   
39            HTML/CSS;Java;JavaScript;Python                   11   
45                        HTML/CSS;JavaScript                   17   
48     HTML/CSS;Java;JavaScript;Kotlin;Python                    8   
49                        HTML/CSS;JavaScript                  118   
...                                       ...                  ...   
51443                 HTML/CSS;JavaScript;PHP                    5   
51470                        Dart;Java;Kotlin                    7   
51482                             Java;Kotlin                   38   
51494                       HTML/CSS;Java;PHP                    3   
51545                C#;JavaScript;Python;SQL                   14   

          ExpLevel   WorkLevel  
11          junior   full-time  
39     experienced  workaholic  
45     experienced   full-time  
48          junior   full-time  
49          junior   full-time  
...            ...         ...  
51443       junior   full-time  
51470  experienced   full-time  
51482  experienced   full-time  
51494  experienced   full-time  
51545  experienced   full-time  

[3749 rows x 11 columns]

In [70]:
# Grouping the data by Country and WorkLevel - return series
grbCountries = (
    df
    .groupby(["Country", "WorkLevel"])
    [["WorkLevel"]]
    .count()
)

# Accessing the data for Israel
WorkLevelIsrael = grbCountries.loc['Israel', 'WorkLevel']
workaholic_percentage = s['workaholic'] / WorkLevelIsrael.sum() * 100

workaholic_percentage

0.7827788649706457

In [71]:
# Accessing the data for India
WorkLevelIndia = grbCountries.loc['India', 'WorkLevel']
workaholic_percentage = s['workaholic'] / WorkLevelIsrael.sum() * 100

workaholic_percentage

0.7827788649706457

In [60]:
israel_workaholics = df[df['Country'] == 'Israel']['WorkLevel'].value_counts(normalize=True).get('workaholic', 0) * 100
india_workaholics = df[df['Country'] == 'India']['WorkLevel'].value_counts(normalize=True).get('workaholic', 0) * 100
# ['WorkLevel'].value_counts(normalize=True) 
# calculates the relative frequency (as a proportion) of each work level category.
# .get('workaholic', 0) retrieves the proportion of respondents categorized as "workaholic"; 
# if there are none, it returns 0.
# Multiplying by 100 converts the proportion to a percentage.

print('israel_workaholics -', israel_workaholics.round(2))

print('india_workaholics -', india_workaholics.round(2))

israel_workaholics - 0.78
india_workaholics - 2.69
